### PR TITLE
Require authentication to post a comment and add comment translation options.

### PR DESCRIPTION
**Pull request by basaam0**
portfolio1
portfolio2
portfolio3

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -63,12 +96,32 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-    response.setContentType("text/html;");
-    response.getWriter().println(json);
+    response.setContentType("application/json;");
+    response.getWriter().println(json.toString());
+  }
+
+  /** 
+   * Returns the name of the user with id, or null if a name has not been set. 
+   */
+  private String getUserName(String id) {
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:131]: You could consider returning an Optional<String> as it removes ambiguity about a reference's nullability (although this is fine for a private method).

File: portfolio/src/main/java/com/google/sps/servlets/NameServlet.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    if (isValidName(name)) {
+      upsertUserInfo(name);
+    }
+  }
+  
+  /**
+   * Checks if the name is composed of only letters, digits, underscores,
+   * and spaces and is between three and twenty characters in length.
+   */
+  private boolean isValidName(String name) {
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/NameServlet.java:57]: If this method doesn't actually require any instance state, you can make this a static method (makes it clear that it does not).

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -46,29 +59,89 @@
...
-      String commentText = (String) entity.getProperty("commentText");
-      long timestamp = (long) entity.getProperty("timestamp");
+    // Check if there are comments to translate.
+    if (results.isEmpty()) {
+      json.add("comments", new JsonArray());
+    } else {
+      // Construct a stream of comment texts from the queried entities.
+      Stream<String> commentTexts = results.stream().map((entity) -> {
+        String commentText = (String) entity.getProperty("commentText");
+        return commentText;
```
Comment (bryyeh) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:101]: If the function within map() is only one line, you don't need brackets or the return statement. It could be shortened to:

entity -> (String) entity.getProperty("commentText")

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -46,29 +59,89 @@
...
+      String urlToRedirectToAfterUserLogsIn = "/";
+      String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+      json.addProperty("loginUrl", loginUrl);
+    }
+    
     // Query up to maxComments comment entities from Datastore with the user's specified sorting option.
+    int maxComments = getMaxCommentsToReturn(request);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     Query query = createCommentQuery(request);
     List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(maxComments));
```
Comment (bryyeh) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:92]: Rename to "entities". Now that the code is getting more complex, "results" is a bit generic.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -46,29 +59,89 @@
...
+      json.add("comments", commentsJsonElement);
+    }
 
-      return new Comment(id, author, commentText, timestamp);
-    }).collect(Collectors.toList());
+    // Send the JSON as the response.
+    response.setContentType("application/json; charset=UTF-8");
+    response.setCharacterEncoding("UTF-8");
+    response.getWriter().println(json.toString());
+  }
```
Comment (bryyeh) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:126]: This method is getting a bit large. Is there a good way to split it up? Perhaps split out the login vs comment logic?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -46,29 +59,89 @@
...
+
+      // Construct a list of comments from the queried entities and translated comment texts.
+      List<Comment> comments = 
+          Streams.zip(results.stream(), translatedCommentTexts, (entity, translatedCommentText) -> {
+            long id = entity.getKey().getId();
+            String author = (String) entity.getProperty("author");
+            long timestamp = (long) entity.getProperty("timestamp");
+
+            return new Comment(id, author, translatedCommentText, timestamp);
+          }).collect(Collectors.toList());
```
Comment (bryyeh) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:115]: Could be placed in a helper method.

File: portfolio/src/main/java/com/google/sps/servlets/LoginStatusServlet.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    
+    // Send the JSON as the response.
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  /**
+   * Converts a boolean value into a JSON string using the Gson library.
+   */
+  private String convertToJson(boolean value) {
```
Comment (bryyeh) [inline portfolio/src/main/java/com/google/sps/servlets/LoginStatusServlet.java:47]: If this method doesn't actually require any instance state, you can make this a static method.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -116,25 +189,42 @@ private Query createCommentQuery(HttpServletRequest request) {
...
+        Translate.TranslateOption.format("text"));
+
+    // Map the translations to a stream of strings.
+    return translations.stream().map(Translation::getTranslatedText);
+  }
+
+  /**
+   * Converts a List into a JsonElement using the Gson library.
+   */
+  private JsonElement convertToJsonElement(List list) {
```
Comment (bryyeh) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:212]: If this method doesn't actually require any instance state, you can make this a static method.

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -119,18 +125,85 @@ function updateColorTheme(backgroundTheme) {
...
+    updateNameForm.style.display = commentsForm.style.display = 'block';
+
+    // Populate the placeholder in the update name form with the stored name.
+    const userNameInput = document.getElementById('user-name-input');
+    userNameInput.placeholder = json.name;
+
+    // Display the user's name.
+    let html = `<p>You are currently logged in as ${json.name}`;
+
+    // Avoid displaying the email twice if it is the same as the name (which is the default).
```
Comment (bryyeh) [inline portfolio/src/main/webapp/script.js:184]: Nice haha

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 10]
```diff
@@ -46,29 +59,89 @@
...
+    if (userService.isUserLoggedIn()) {
+      String name = getUserName(userService.getCurrentUser().getUserId());
+
+      // If the user has logged in for the first time, set their name as their Google account nickname.
+      if (name == null) {
+        name = userService.getCurrentUser().getNickname();
+        NameServlet.upsertUserInfo(name);
+      }
+
+      String urlToRedirectToAfterUserLogsOut = "/";
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:74]: nit: since the redirect url is actually the same for log out and log in, maybe just create a single variable `redirectUrl = "/";` or `urlToRedirectTo = "/";`

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 11]
```diff
@@ -46,29 +59,89 @@
...
+        String commentText = (String) entity.getProperty("commentText");
+        return commentText;
+      });
+
+      String languageCode = request.getParameter("language-code");
+      Stream<String> translatedCommentTexts = translateComments(commentTexts, languageCode);
+
+      // Construct a list of comments from the queried entities and translated comment texts.
+      List<Comment> comments = 
+          Streams.zip(results.stream(), translatedCommentTexts, (entity, translatedCommentText) -> {
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:109]: Nice use of zip! My only concern was making sure the the results of the translation call is returned in the same order which it seems that it should be based on the documentation at [1]:

```
Returns:
a list of objects containing information on the language translation, one for each provided text, in order
```

[1] https://googleapis.dev/java/google-cloud-translate/latest/com/google/cloud/translate/Translate.html#translate-java.util.List-com.google.cloud.translate.Translate.TranslateOption...-

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 12]
```diff
@@ -15,26 +15,39 @@
...
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 /** 
- * Servlet with a GET handler that loads a list of comments from Datastore and 
+ * Servlet with a GET handler that loads a list of comments from Datastore, checks
+ * whether the user is logged in, and sends back either the user's email, name, and
+ * a link to logout if they are logged in or a link to login if they are not; and
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:50]: Now that there is a lot more functionality maybe some of these details / description could be added to the respective methods doGet(), doPost()?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 13]
```diff
@@ -46,29 +59,89 @@
...
-    int maxComments = getMaxCommentsToReturn(request);
-
+    JsonObject json = new JsonObject();
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (userService.isUserLoggedIn()) {
+      String name = getUserName(userService.getCurrentUser().getUserId());
+
+      // If the user has logged in for the first time, set their name as their Google account nickname.
+      if (name == null) {
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:69]: If the user updates their name, how would this be handled? I'm wondering if we need to store our own id->name mapping and managing user names. UserService may have their own privacy requirements and it may not be clear to users also that this website would be storing a copy of the name info in its own datastore.

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -119,18 +125,85 @@ function updateColorTheme(backgroundTheme) {
...
 
 /**
- * Fetches the list of comments from the server and adds them to the DOM.
+ * Displays either the form to post a comment if the user is logged in or a link to login if they are not
+ * and displays the list of comments. 
  */
-async function getComments() {
+async function loadCommentsSection() {
+  // Set the cursor to the loading icon.
+  document.body.style.cursor = 'wait';
```
Comment (acheng1) [inline portfolio/src/main/webapp/script.js:137]: Nice touch to the website. Like the attention to detail.

File: portfolio/src/main/webapp/script.js [hunk 15]
```diff
@@ -119,18 +125,85 @@ function updateColorTheme(backgroundTheme) {
...
+
+  const response = await fetch(`/data?${params.toString()}`);
+  const json = await response.json();
+
+  // Load the form to post comments and the list of comments.
+  await Promise.all([
+    getCommentsForm(json), 
+    getComments(json.comments),
+  ]);
+  // Restore the cursor after the comments sectuib has loaded.
```
Comment (acheng1) [inline portfolio/src/main/webapp/script.js:156]: section?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 13 above)
Comment (acheng1) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:69]: Ah, as I read more of the code I see there's a way to update your stored name. You are just seeding the name from the UserService. I was curious about what happens if they updated their nickname on the UserService. I'm still curious if we want to maintain different identities?

Comment (basaam0) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:69]: My intention was to allow users to have a display name specific to the portfolio and just use the nickname from UserService as an initial value before the users update it. So if the nickname in UserService changes after a user first logs in, it should not have any effect. Does that answer your question?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 16]
```diff
@@ -44,31 +56,101 @@
   private static final Logger LOGGER = Logger.getLogger(DataServlet.class.getName());
   private static final int DEFAULT_MAX_COMMENTS = 10;
 
+  /**
+   * Loads the list of comments from Datastore and applies the requested limit, sort,
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:60]: Nice description!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 17]
```diff
@@ -44,31 +56,101 @@
...
+            .setFilter(new Query.FilterPredicate("id", Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return Optional.empty();
+    }
+
+    String name = (String) entity.getProperty("name");
+    return Optional.of(name);
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:153]: You can shorten this to single line using `Optional.ofNullable()` [1] with `.map` [2]

```
return Optional.ofNullable(entity).map(...)
```

[1] https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#ofNullable-T-
[2] https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#map-java.util.function.Function-

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 18]
```diff
@@ -116,30 +198,73 @@ private Query createCommentQuery(HttpServletRequest request) {
...
     
-    // Use the default author "Anonymous" if none is provided.
-    if (author.isEmpty()) {
-      author = Comment.DEFAULT_AUTHOR;
+    String userId = userService.getCurrentUser().getUserId();
+    Optional<String> author = getUserName(userId);
+
+    if (!author.isPresent()) {
+      String email = userService.getCurrentUser().getEmail(); 
+      LOGGER.severe("User <" + email + "> does not have an associated name.");
```
Comment (acheng1) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:257]: nice!
